edb/edgeql-parser/edgeql-parser-python/src/parser.rs

```diff
@@ -105,46 +110,69 @@ fn downcast_tokens<'a>(
     Ok(buf)
 }
 
-pub fn cache_spec(py: Python, py_spec: &PyObject) -> PyResult<PyNone> {
-    if PARSER_SPECS.get().is_some() {
-        return Ok(PyNone);
+fn get_spec(
+    py: Python,
+    spec_filepath: &str,
+) -> Result<&'static (parser::Spec, PyObject), cpython::PyErr> {
+    if let Some(x) = PARSER_SPECS.get() {
+        return Ok(x);
     }
 
-    let x = load_spec(py, py_spec)?;
-    PARSER_SPECS.set(x).ok();
+    // Try to load from file
+    let spec = load_spec(spec_filepath);
+    let productions = load_productions(py, &spec)?;
+
+    PARSER_SPECS.set((spec, productions)).ok();
+    Ok(PARSER_SPECS.get().unwrap())
+}
+
+/// Loads the grammar specification from file and caches it in memory.
+pub fn preload_spec(py: Python, spec_filepath: &PyString) -> PyResult<PyNone> {
+    get_spec(py, &spec_filepath.to_string(py)?)?;
     Ok(PyNone)
 }
 
-fn get_spec(py: Python<'_>) -> Result<&(parser::Spec, PyObject), cpython::PyErr> {
-    if let Some(x) = PARSER_SPECS.get() {
-        return Ok(x);
-    }
+/// Serialize the grammar specification and write it to a file.
+///
+/// Called from setup.py.
+pub fn save_spec(py: Python, py_spec: &PyObject, dst: &PyString) -> PyResult<PyNone> {
+    let spec = spec_from_python(py, py_spec)?;
+    let spec_serialized = bitcode::serialize(&spec).unwrap();
 
-    let parsing_mod = py.import("edb.common.parsing")?;
-    let load_parser_spec = parsing_mod.get(py, "load_parser_spec")?;
+    let dst = dst.to_string(py)?.to_string();
 
-    let grammar_name = "edb.edgeql.parser.grammar.start";
-    let grammar_mod = py.import(grammar_name)?;
-    let py_spec = load_parser_spec.call(py, (grammar_mod,), None)?;
+    std::fs::write(dst, spec_serialized).ok().unwrap();
+    Ok(PyNone)
+}
 
-    let x = load_spec(py, &py_spec)?;
+fn load_spec(spec_filepath: &str) -> parser::Spec {
+    let bytes = std::fs::read(spec_filepath).ok().unwrap();
 
-    PARSER_SPECS.set(x).ok();
-    Ok(PARSER_SPECS.get().unwrap())
+    bitcode::deserialize::<parser::SpecSerializable>(&bytes)
```

Comment:
How big are the savings from storing the serialized spec as bitcode instead of just saving the json?
Mostly just curious.

Reply:
bitcode is 47% smaller:
```
[edb/edgeql-parser/edgeql-parser-python/src/parser.rs:160] spec_json.len() = 9473372
[edb/edgeql-parser/edgeql-parser-python/src/parser.rs:144] spec_serialized.len() = 4470568
```